[BUGFIX] Le InMemoryTemporaryStorage ne suit pas la TemporaryStorage API et est utilisé à tort dans les tests d'intégration (PIX-12551)

lego-technix opened the pull request:
## :unicorn: Problème

Le `InMemoryTemporaryStorage` ne suit pas la `TemporaryStorage` API et est utilisé à tort dans les tests d'intégration.

Une conséquence de cela est que le `PoleEmploiOidcAuthenticationService` et le `FwbOidcAuthenticationService` sont testés avec une mauvaise implémentation et leurs comportements sont dysfonctionnels concernant la méthode `getRedirectLogoutUrl` qui ne fournit pas un`'id_token_hint`.

## :robot: Proposition

1. Faire suivre la `TemporaryStorage` API au `InMemoryTemporaryStorage` en rendant les méthodes du second asynchrones comme le sont celles du premier
2. Pour les tests d'intégration utiliser le `RedisTemporaryStorage` à la place du `InMemoryTemporaryStorage`, car les tests d'intégration servent à tester les composants dans le cadre d'un environnement où l'infrastructure est la plus proche possible de celle de la production

## :rainbow: Remarques

Il faudrait également que les tests d'intégration mettant en jeu le `LearningContentCache` utilisent _Redis_ au lieu d'utiliser uniquement le `InMemoryCache`.

Pour cela il faudrait mettre dans `api/src/shared/config.js` : 
```javascript
config.caching.redisUrl = process.env.TEST_REDIS_URL;
```

Et dans `api/tests/acceptance/application/cache/cache-controller_test.js` il faudrait utiliser `config.caching.redisUrl ` au lieu d'utiliser `process.env.TEST_REDIS_URL` directement.

Mais cette correction s'est avérée difficile et on n'a pas réussi à en venir à bout dans un temps limité et nous avons abandonné cette correction qui n'est pas dans le scope de l'équipe Accès.

## :100: Pour tester

1. Vérifier que la CI s'exécute sans erreur
2. Vérifier aussi en local que les tests s'exécutent sans erreur
   * Vérifier en local que les tests unitaires utilisent le `InMemoryTemporaryStorage` (pour cela on peut arrêter le conteneur de Redis – `docker stop <container_id>` – et constater que les tests unitaires continuent de fonctionner sans erreur)
   * Vérifier en local que les tests d'intégration et d'acceptance utilisent le `RedisTemporaryStorage` (pour cela on peut arrêter le conteneur de Redis – `docker stop <container_id>` – et constater que les tests d'intégration sont en erreur)
3. Vérifier que la déconnexion pour France Travail et la FWB déconnecte bien l'utilisateur du SSO du fournisseur d'identité

lego-technix committed:
fix(api): make InMemoryTemporaryStorage implement TemporaryStorage API

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Eric Lim <eric.lim@pix.fr>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>
Co-authored-by: Marianne Bost <marianne.bost@pix.fr>

lego-technix committed:
fix(api): fix POLE_EMPLOI and FWB TemporaryStorage API usage

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Eric Lim <eric.lim@pix.fr>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>
Co-authored-by: Marianne Bost <marianne.bost@pix.fr>

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr8954.review.pix.fr)
  * [App (.org)](https://app-pr8954.review.pix.org)
  * [Orga (.fr)](https://orga-pr8954.review.pix.fr)
  * [Orga (.org)](https://orga-pr8954.review.pix.org)
  * [Certif (.fr)](https://certif-pr8954.review.pix.fr)
  * [Certif (.org)](https://certif-pr8954.review.pix.org)
  * [Junior](https://junior-pr8954.review.pix.fr)
  * [Admin](https://admin-pr8954.review.pix.fr)
  * [API](https://api-pr8954.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr8954.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr8954/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr8954/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr8954/environment)

lego-technix committed:
fix(api): don't use InMemoryTemporaryStorage for integration tests

rather use RedisTemporaryStorage instead, which is the true infrastructure
component used in production.

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Eric Lim <eric.lim@pix.fr>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@

lego-technix committed:
refactor(api): clearer redisUrl variable name

wrt the REDIS_URL environment variable.

lego-technix committed:
refactor(api): remove unused config.redis.database

EmmanuelleBonnemay commented on api/src/shared/config.js:
Cette configuration `database : 1` a-t-elle toujours été inutile ou est-elle rendue inutile par les modifications de cette PR ?

lego-technix replied:
Cette configuration `database : 1` n'a pas été rendue inutile par les modifications de cette PR. Je ne sais pas de quand elle date et pourquoi, par contre je sais qu'elle n'est utilisée nulle par à présent.

EmmanuelleBonnemay commented:
En ce qui concerne les tests unitaires, j'ai vérifié de la même manière que pour les tests d'intégration (arrêter le conteneur Redis et vérifier que les tests unitaires n'ont pas besoin de ce conteneur pour s'exécuter). Est-ce que cela suffit? Il n'y a pas de log spécifique au `InMemoryTemporaryStorage` en console :
```
> pix-api@4.146.0 test:api:unit
> TEST_DATABASE_URL=postgres://should.not.reach.db.in.unit.tests TEST_REDIS_URL= npm run test:api:path -- 'tests/**/unit/**/*test.js'
```

lego-technix commented:
> En ce qui concerne les tests unitaires, j'ai vérifié de la même manière que pour les tests d'intégration (arrêter le conteneur Redis et vérifier que les tests unitaires n'ont pas besoin de ce conteneur pour s'exécuter). Est-ce que cela suffit? Il n'y a pas de log spécifique au `InMemoryTemporaryStorage` en console :

Tout à fait. Il n'y a pas de logs spécifiques au `InMemoryTemporaryStorage` en console. Donc on peut soit tester en arrêtant/redémarrant les conteneurs Docker, soit en ajoutant soi-même des logs dans le `InMemoryTemporaryStorage` et le `RedisTemporaryStorage`.

EmmanuelleBonnemay reviewed (approved):
Lu et testé fonctionnellement en local avec succès sur Firefox.
